src/decoder/lattice-incremental-decoder.h

```diff
@@ -0,0 +1,655 @@
+
+  // This function takes a singly linked list of tokens for a single frame, and
+  // outputs a list of them in topological order (it will crash if no such order
+  // can be found, which will typically be due to decoding graphs with epsilon
+  // cycles, which are not allowed).  Note: the output list may contain NULLs,
+  // which the caller should pass over; it just happens to be more efficient for
+  // the algorithm to output a list that contains NULLs.
+  static void TopSortTokens(Token *tok_list, std::vector<Token *> *topsorted_list);
+
+  void ClearActiveTokens();
+
+  // The following part is specifically designed for incremental determinization
+  // This function is modified from LatticeFasterDecoderTpl::GetRawLattice()
+  // and specific design for step 1 of incremental determinization
+  // introduced before above GetLattice()
+  // It does the same thing as GetRawLattice in lattice-faster-decoder.cc except:
+  //
+  // i) it creates a initial state, and connect
+  // each token in the first frame of this chunk to the initial state
+  // by one or more arcs with a state_label correponding to the pre-final state w.r.t
+  // this token(the pre-final state is appended in the last chunk) as its olabel
+  // ii) it creates a final state, and connect
+  // all the tokens in the last frame of this chunk to the final state
+  // by an arc with a per-token token_label as its olabel
+  // `frame_begin` and `frame_end` are the first and last frame of this chunk
+  // if `create_initial_state` == false, we will not create initial state and
+  // the corresponding initial arcs. Similar for `create_final_state`
+  // In incremental GetLattice, we do not create the initial state in
+  // the first chunk, and we do not create the final state in the last chunk
+  bool GetIncrementalRawLattice(Lattice *ofst, bool use_final_probs,
+                                int32 frame_begin, int32 frame_end,
+                                bool create_initial_state, bool create_final_state);
+  // Get the number of tokens in each frame
+  // It is useful, e.g. in using config_.determinize_max_active
+  int32 GetNumToksForFrame(int32 frame);
+
+  // The incremental lattice determinizer to take care of step 2-4
+  LatticeIncrementalDeterminizer<FST> determinizer_;
+  int32 last_get_lattice_frame_; // the last time we call GetLattice
+  // a map from Token to its token_label
+  unordered_map<Token *, int32> token_label_map_;
+  // we allocate a unique id for each Token
+  int32 token_label_available_idx_;
+  // We keep cost_offset for each token_label (Token) in final arcs. We need them to
+  // guide determinization
+  // We cancel them after determinization
+  unordered_map<int32, BaseFloat> token_label_final_cost_;
+  KALDI_DISALLOW_COPY_AND_ASSIGN(LatticeIncrementalDecoderTpl);
+};
+
+typedef LatticeIncrementalDecoderTpl<fst::StdFst, decoder::StdToken>
+    LatticeIncrementalDecoder;
+
+// This class is designed for step 2-4 and part of step 1 of incremental
```

Comment:
Where are these steps defined?  You can't just refer to concepts like this without explaining where they came from.
If this notation does not appear in the paper or is not clearly explained here, we shouldn't include it.

Reply:
I define steps before the definition of GetLattice() in line 217

Reply:
You can't assume people will have read what goes above, unless it's very closely above.  You have to make it navigable, e.g. tell people how to find the definitions of those terms, when you use those terms.